service/history/replicationTaskProcessor.go

```diff
@@ -0,0 +1,285 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"time"
+
+	h "github.com/uber/cadence/.gen/go/history"
+	r "github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/service/worker/replicator"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+const (
+	dropSyncShardTaskTimeThreshold = 10 * time.Minute
+	replicationTimeout             = 30 * time.Second
+)
+
+var (
+	// ErrUnknownReplicationTask is the error to indicate unknown replication task type
+	ErrUnknownReplicationTask = &shared.BadRequestError{Message: "unknown replication task"}
+)
+
+type (
+	replicationTaskProcessor struct {
+		shard            ShardContext
+		readLevel        int64
+		historyEngine    Engine
+		sourceCluster    string
+		domainReplicator replicator.DomainReplicator
+		metricsClient    metrics.Client
+		logger           log.Logger
+
+		requestChan chan<- *request
+		done        chan struct{}
+	}
+
+	request struct {
+		token    *r.ReplicationToken
+		respChan chan<- *r.ReplicationTasksInfo
+	}
+)
+
+func newReplicationTaskProcessor(
+	shard ShardContext,
+	historyEngine Engine,
+	domainReplicator replicator.DomainReplicator,
+	metricsClient metrics.Client,
+	replicationTaskFetcher *replicationTaskFetcher,
+) *replicationTaskProcessor {
+	return &replicationTaskProcessor{
+		shard:            shard,
+		historyEngine:    historyEngine,
+		sourceCluster:    replicationTaskFetcher.GetSourceCluster(),
+		domainReplicator: domainReplicator,
+		metricsClient:    metricsClient,
+		logger:           shard.GetLogger(),
+		requestChan:      replicationTaskFetcher.GetRequestChan(),
+		done:             make(chan struct{}),
+	}
+}
+
+func (p *replicationTaskProcessor) Start() {
+	go func() {
+		p.readLevel = p.shard.GetClusterReplicationLevel(p.sourceCluster)
+		scope := p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster))
+
+		for {
```

Comment:
Do we need any rating control for this loop?

Reply:
@wxing1292 @samarabbas what do you think?

Reply:
replicationTaskFetcher already has rate limiting, so i guess it is fine. plz add a comment.